File: src/live/navigation.ts
export type Point = [number, number];
export type Arena = { arena: number; clearance: number; obstacles: { x: number; y: number; hx: number; hy: number; height: number }[] };
export const distance = (a: number[], b: number[]) => Math.hypot(a[0]-b[0],a[1]-b[1]);

export function clear(point: number[], arena: Arena, margin = arena.clearance) {
  if (Math.max(Math.abs(point[0]),Math.abs(point[1])) > arena.arena-margin) return false;
  return arena.obstacles.every(o => Math.hypot(Math.max(Math.abs(point[0]-o.x)-o.hx,0),Math.max(Math.abs(point[1]-o.y)-o.hy,0)) > margin);
}

export function segmentClear(a: number[], b: number[], arena: Arena, margin = arena.clearance) {
  const samples = Math.max(1,Math.ceil(distance(a,b)/.025));
  for (let i=0; i<=samples; i++) if (!clear([a[0]+(b[0]-a[0])*i/samples,a[1]+(b[1]-a[1])*i/samples],arena,margin)) return false;
  return true;
}

export function planPath(start: Point, goal: Point, arena: Arena): Point[] {
  if (!goal.every(Number.isFinite) || !clear(start,arena) || !clear(goal,arena)) throw new Error('Choose clear floor away from obstacles and walls.');
  if (segmentClear(start,goal,arena)) return [start,goal];
  const cell = (p: Point): Point => [Math.round((p[0]+arena.arena)/.1),Math.round((p[1]+arena.arena)/.1)];
  const point = (c: Point): Point => [c[0]*.1-arena.arena,c[1]*.1-arena.arena];
  const key = (c: Point) => c.join(',');
  const first=cell(start), last=cell(goal);
  if (!segmentClear(start,point(first),arena) || !segmentClear(point(last),goal,arena)) throw new Error('Move the destination slightly away from the clearance boundary.');
  const open = [{cell:first,priority:0}];
  const costs = new Map([[key(first),0]]), parent = new Map<string,Point>();
  while (open.length) {
    open.sort((a,b) => a.priority-b.priority);
    const current = open.shift()!.cell;
    if (key(current) === key(last)) break;
    for (let dx=-1; dx<=1; dx++) for (let dy=-1; dy<=1; dy++) {
      if (!dx && !dy) continue;
      const next: Point = [current[0]+dx,current[1]+dy];
      if (!segmentClear(point(current),point(next),arena)) continue;
      const cost = costs.get(key(current))! + Math.hypot(dx,dy)*.1;
      if (cost < (costs.get(key(next)) ?? Infinity)) {
        costs.set(key(next),cost); parent.set(key(next),current);
        open.push({cell:next,priority:cost+distance(point(next),point(last))});
      }
    }
  }
  if (!costs.has(key(last))) throw new Error('No route with enough room for all four feet.');
  const cells = [last];
  while (key(cells.at(-1)!) !== key(first)) cells.push(parent.get(key(cells.at(-1)!))!);
  const points: Point[] = [start,...cells.reverse().map(point),goal];
  const result = [start];
  for (let i=0; i<points.length-1;) {
    let j=points.length-1;
    while (j>i+1 && !segmentClear(points[i],points[j],arena)) j--;
    result.push(points[j]); i=j;
  }
  return result;
}

File: simulation/navigation.py
"""Known-map A* navigation for the slow crawl. World coordinates are metres."""
import heapq
import math
import numpy as np
import mujoco
from .model import model_xml, HOME_HEIGHT, FOOT_RADIUS, LENGTH

ARENA = 2.4
ROBOT_CLEARANCE = .65
OBSTACLES = [
    {"x": .95, "y": .35, "hx": .18, "hy": .28, "height": .45},
    {"x": -.85, "y": .85, "hx": .22, "hy": .18, "height": .32},
    {"x": .30, "y": -1.10, "hx": .38, "hy": .14, "height": .30},
]


def clear(point, margin=ROBOT_CLEARANCE):
    x, y = point
    if max(abs(x),abs(y)) > ARENA-margin:
        return False
    for obstacle in OBSTACLES:
        dx = max(abs(x-obstacle["x"])-obstacle["hx"],0.)
        dy = max(abs(y-obstacle["y"])-obstacle["hy"],0.)
        if math.hypot(dx,dy) <= margin:
            return False
    return True


def segment_clear(start, end, margin=ROBOT_CLEARANCE):
    distance = np.linalg.norm(np.array(end)-start)
    return all(clear(np.array(start)+(np.array(end)-start)*t,margin) for t in np.linspace(0,1,max(2,math.ceil(distance/.025)+1)))


def plan_path(start, goal):
    """8-connected A*, then shortcut only collision-free line segments."""
    start, goal = np.array(start), np.array(goal)
    if not clear(start) or not clear(goal):
        raise ValueError("Choose a point in the clear area, away from obstacles and walls.")
    if segment_clear(start, goal):
        return [start.tolist(), goal.tolist()]
    resolution = .10
    origin = -ARENA
    to_cell = lambda p: tuple(np.round((p-origin)/resolution).astype(int))
    to_point = lambda c: np.array(c)*resolution + origin
    first, last = to_cell(start), to_cell(goal)
    if not segment_clear(start,to_point(first)) or not segment_clear(to_point(last),goal):
        raise ValueError("Destination is too close to the clearance boundary; move it slightly.")
    queue, costs, parent = [(0., first)], {first: 0.}, {}
    while queue:
        _, cell = heapq.heappop(queue)
        if cell == last:
            break
        for dx in (-1,0,1):
            for dy in (-1,0,1):
                if dx == dy == 0:
                    continue
                candidate = (cell[0]+dx, cell[1]+dy)
                p = to_point(candidate)
                if not clear(p) or not segment_clear(to_point(cell),p):
                    continue
                cost = costs[cell]+math.hypot(dx,dy)*resolution
                if cost < costs.get(candidate, float("inf")):
                    costs[candidate], parent[candidate] = cost, cell
                    heapq.heappush(queue,(cost+np.linalg.norm(p-to_point(last)),candidate))
    if last not in costs:
        raise ValueError("No route with enough clearance for all four feet.")
    cells = [last]
    while cells[-1] != first:
        cells.append(parent[cells[-1]])
    points = [start.tolist()] + [to_point(c).tolist() for c in reversed(cells)] + [goal.tolist()]
    simplified = [points[0]]
    i = 0
    while i < len(points)-1:
        j = len(points)-1
        while j > i+1 and not segment_clear(points[i],points[j]):
            j -= 1
        simplified.append(points[j])
        i = j
    return simplified


def make_arena():
    obstacles = []
    for i, o in enumerate(OBSTACLES):
        obstacles.append(f'<geom name="obstacle_{i}" type="box" pos="{o["x"]} {o["y"]} {o["height"]/2}" size="{o["hx"]} {o["hy"]} {o["height"]/2}" rgba=".28 .37 .40 1"/>')
    for x,y,sx,sy in ((ARENA+.05,0,.05,ARENA+.1),(-ARENA-.05,0,.05,ARENA+.1),(0,ARENA+.05,ARENA,.05),(0,-ARENA-.05,ARENA,.05)):
        obstacles.append(f'<geom type="box" pos="{x} {y} .12" size="{sx} {sy} .12" rgba=".23 .32 .35 1"/>')
    xml = model_xml().replace('<worldbody>','<worldbody>'+''.join(obstacles))
    model = mujoco.MjModel.from_xml_string(xml)
    data = mujoco.MjData(model)
    bend = np.arccos((HOME_HEIGHT-FOOT_RADIUS)/(2*LENGTH))
    data.qpos[7:] = np.tile([0,-bend,2*bend],4)
    mujoco.mj_forward(model,data)
    return model, data
